File: GriffBallJava/GriffServer/src/GriffBall.java
import java.net.Socket;
import java.util.Map;
import java.util.TreeMap;

public class GriffBall {
    private static Map<Integer, Player> players = new TreeMap<>();

    public void addPlayer(int playerId, String name) {
        Player player = new Player(playerId, name);
        players.put(playerId, player);

        if (players.size() == 1) {
            giveBall(playerId);
        }
    }

    public void removePlayer(int playerid) {
        players.remove(playerid);
    }

    public void giveBall(int playerId) {
        //checkCorrectState();
        String outputBuilder = "";
        String secondPlayer = "";

        for (Player player : players.values()) {
            if (player.hasBall)
                outputBuilder += "Player_" + player.getId() + "(" + player.getName() + ") Has thrown the ball to ";
            if (player.getId() != playerId) {
                player.hasBall = false;
            } else {
                player.hasBall = true;
                secondPlayer = "Player_" + player.getId() + "(" + player.getName() + ")";
            }
        }

        outputBuilder = outputBuilder + secondPlayer;

        System.out.println(outputBuilder);
        System.out.println("Player " + playerId + " Has the ball");
    }

    public synchronized void checkCorrectState() {
        // This function checks that the game state is correct and fixes the game if it is not
        int ballCount = 0;
        for (Player player : players.values()) {
            if (player.hasBall) {
                ballCount++;
            }
        }
        if (ballCount != 1) {
            System.out.println("Ball Exploded!");
            for (Player player : players.values()) {
                giveBall(player.getId());
                break;
            }
        }
    }

    public synchronized String whoHasBall() {
        String result = "";  // [Player_id(name) id]
        checkCorrectState();

        for (Player player : players.values()) {
            if (player.hasBall) {
                result = "Player_" + player.getId() + "(" + player.getName() + ") " + player.getId();
                break;
            }
        }

        return result;
    }

    public String getPlayersString() {
        String stringBuilder = "";

        for (Player player : players.values()) {
            stringBuilder += "Player_" + player.getId() + "(" + player.getName() + ") ";
        }
        stringBuilder = stringBuilder.trim();

        return stringBuilder;
    }

    public void printPlayers() {
        String stringBuilder = ">Players:" + players.size();

        for (Player player : players.values()) {
            stringBuilder += "\n" + player.getId() + ":" + player.getName();
        }

        System.out.println(stringBuilder);
    }
}

File: GriffBallServer/GriffBall.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GriffBallServer
{
    class GriffBall
    {
        private static Dictionary<int, Player> players = new Dictionary<int, Player>();
        private static readonly object sync = new object(); // Used to create synchronized methods

        public void addPlayer(int playerId, string name)
        {
            Player player = new Player(playerId, name);
            players.Add(playerId, player);

            if (players.Count == 1)
            {
                giveBall(playerId);
            }
        }

        public void removePlayer(int playerId)
        {
            players.Remove(playerId);
        }

        public void giveBall(int playerId)
        {
            string outputBuilder = "";
            string secondPlayer = "";

            foreach (Player player in players.Values)
            {
                if (player.hasBall)
                    outputBuilder += "Player_" + player.id + "(" + player.name + ") has thrown the ball to ";

                if (player.id != playerId)
                {
                    player.hasBall = false;
                }
                else
                {
                    player.hasBall = true;
                    secondPlayer = "Player_" + player.id + "(" + player.name + ")";
                }
            }

            outputBuilder = outputBuilder + secondPlayer;

            Console.WriteLine(outputBuilder);
            Console.WriteLine("Player " + playerId + " Has the ball");
        }

        public void checkCorrectState()
        {
            lock (sync)
            {
                int ballCount = 0;

                foreach (Player player in players.Values)
                {
                    if (player.hasBall) { ballCount++; }
                }
                if (ballCount != 1)
                {
                    Console.WriteLine("Ball Exploded!");
                    foreach (Player player in players.Values)
                    {
                        giveBall(player.id);
                        break;
                    }
                }
            }
        }

        public String whoHasBall()
        {
            lock (sync)
            {
                String result = "";
                checkCorrectState();

                foreach (Player player in players.Values)
                {
                    if (player.hasBall)
                    {
                        result = "Player_" + player.id + "(" + player.name + ") " + player.id;
                        break;
                    }
                }

                return result;
            }
        }

        public string getPlayersString()
        {
            string stringBuilder = "";

            foreach (Player player in players.Values)
            {
                stringBuilder += "Player_" + player.id + "(" + player.name + ") ";
            }
            stringBuilder = stringBuilder.Trim();

            return stringBuilder;
        }

        public void printPlayers()
        {
            String stringBuilder = ">Players:" + players.Count;

            foreach (Player player in players.Values)
            {
                stringBuilder += "\n" + player.id + ":" + player.name;
            }

            Console.WriteLine(stringBuilder);
        }
    }
}
